### ISSUE TITLE
`kubeadm upgrade apply v1.30.0` loses custom kubelet configuration

### ISSUE THREAD

**Issue by ilia1243** [kubernetes/kubeadm#3049]
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): v1.30.0

**Environment**:
- **Kubernetes version** (use `kubectl version`): v1.30.0
- **Cloud provider or hardware configuration**: 
- **OS** (e.g. from /etc/os-release): Ubuntu 22.04.1 LTS
- **Kernel** (e.g. `uname -a`): 5.15.0-50-generic
- **Container runtime (CRI)** (e.g. containerd, cri-o): containerd=1.6.12-0ubuntu1~22.04.3
- **Container networking plugin (CNI)** (e.g. Calico, Cilium): calico
- **Others**:

## What happened?

`kubeadm upgrade apply v1.30.0` loses custom kubelet (and probably kube-proxy) configuration, including both /var/lib/kubelet/config.yaml and kubelet-config ConfigMap.

This is probably caused by PR https://github.com/kubernetes/kubernetes/pull/123068 after which ComponentConfigs are no longer fetched
https://github.com/kubernetes/kubernetes/blob/v1.30.0/cmd/kubeadm/app/cmd/upgrade/common.go#L102
for the config that is used to upload and write kubelet configs
https://github.com/kubernetes/kubernetes/blob/v1.30.0/cmd/kubeadm/app/cmd/upgrade/apply.go#L216

## What you expected to happen?

Kubelet configuration should be preserved.

## How to reproduce it (as minimally and precisely as possible)?

1. Install the cluster of version v1.29.1
2. `kubeadm upgrade apply v1.30.0` the first control plane.

**Comment by neolit123** [comment]
the correct way to pass per-node kubelet configuration is by using --patches with a kubeletconfiguration target during init/join/upgrade
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#kubelet

--config on upgrade incorrectly supported passing kubeletconfiguration and kubeproxyconfiguration.


> kubeadm upgrade apply v1.30.0 loses custom kubelet (and probably kube-proxy) configuration, including both /var/lib/kubelet/config.yaml and kubelet-config ConfigMap.

just to doub;e check, do you mean that the kubeletconfiguration is no longer downloaded from the cluster kubelet-config CM or just that it's no longer possible to upload a new config to the kubelet-config CM?

**Comment by ilia1243** [comment]
We use --patches, and do not use --config. The problem with custom global configuration, not per-node one.

> just to doub;e check, do you mean that the kubeletconfiguration is no longer downloaded from the cluster kubelet-config CM or just that it's no longer possible to upload a new config to the kubelet-config CM during upgrade?

The kubeletconfiguration is no longer downloaded from the cluster kubelet-config CM.

Thus, during upgrade kubelet-config CM is not taken account, and per-node configs are overwritten with just defaults.

**Comment by neolit123** [comment]
> The kubeletconfiguration is no longer downloaded from the cluster kubelet-config CM.
> Thus, during upgrade kubelet-config CM is not taken account, and per-node configs are overwritten with just defaults.

ok, if true that is a bad regression.
i will test it locally.

**Comment by neolit123** [comment]
confirmed. this is a regression due to the introduction of upgradeconfiguration and highlights that we don't have good tests.
ideally we should have the kubeadm upgrade apply phases one day and consolidate the whole "apply" upgrade logic.

> This is probably caused by PR https://github.com/kubernetes/kubernetes/pull/123068 after which ComponentConfigs are no longer fetched
https://github.com/kubernetes/kubernetes/blob/v1.30.0/cmd/kubeadm/app/cmd/upgrade/common.go#L102
for the config that is used to upload and write kubelet configs
https://github.com/kubernetes/kubernetes/blob/v1.30.0/cmd/kubeadm/app/cmd/upgrade/apply.go#L216

@calvin0327 @SataQiu @pacoxu 

i will try to fix this.

**Comment by neolit123** [comment]
PR
https://github.com/kubernetes/kubernetes/pull/124480

once we agree on the 1.31 fix this can be backported and added to 1.30.1.

**Comment by neolit123** [comment]
FYI, for "upgrade node" this is not a problem because it has:

```
initCfg, err := configutil.FetchInitConfigurationFromCluster(client, nil, "upgrade", !isControlPlaneNode, false)
```

i.e. last arg is false, as in, don't skip component config download.

**Comment by neolit123** [comment]
fix will be available in 1.31
https://github.com/kubernetes/kubernetes/pull/124497

**Comment by neolit123** [comment]
1.30.1 is out with the fix
